### PR TITLE
add Binomial distribution

### LINKED ISSUE #42: Implement Binomial distribution

### PR DESCRIPTION
**Pull request by alicanb**
closes #42

### CODE REVIEW

File: torch/distributions/binomial.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    Creates a Binomial distribution parameterized by `total_count` and
+    either `probs` or `logits` (but not both).
+
+    -   :meth:`sample` requires a single shared `total_count` for all
+        parameters and samples.
+
+    Example::
+
+        >>> m = Binomial(100, torch.Tensor([0 , .2, .8, 1]))
+        >>> x = m.sample()
```
Comment (fritzo) [inline torch/distributions/binomial.py]: Could you update these with a plausible sample? it should look like `[0, 22, 74, 100]`.

File: torch/distributions/binomial.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    @lazy_property
+    def logits(self):
+        return probs_to_logits(self.probs)
+
+    @lazy_property
+    def probs(self):
+        return self.logits.exp()
+
+    def sample(self, sample_shape=torch.Size()):
+        shape = (self.total_count,) + self._extended_shape(sample_shape)
```
Comment (fritzo) [inline torch/distributions/binomial.py]: nit: I believe it's slightly cheaper to sum over the rightmost dimension than the leftmost dimension due to memory locality.

File: torch/distributions/binomial.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    def log_prob(self, value):
+        self._validate_log_prob_arg(value)
+        probs = clamp_probs(self.probs)
+        log_factorial_n = math.lgamma(self.total_count + 1)
+        log_factorial_k = torch.lgamma(value + 1)
+        log_factorial_nmk = torch.lgamma(self.total_count - value + 1)
+        return (log_factorial_n - log_factorial_k - log_factorial_nmk +
+                value * self.logits + (self.total_count - value) * torch.log1p(-probs))
+
+    def entropy(self):
```
Comment (fritzo) [inline torch/distributions/binomial.py]: This is only approximate, right? I feel like we'd be better off leaving it not implemented than implementing an approximate entropy.

Also, we should adapt the TestKL tests to entropy tests to catch things like this.

**Comment by alicanb** [comment]
Thanks @fritzo for the review

**Comment by alicanb** [comment]
Sending upstream

**Pull request closed without merging**